File: app/src/main/java/se/gustavkarlsson/aurora_notifier/android/gui/activities/main/fragments/aurora_factors/AuroraFactorView.java
package se.gustavkarlsson.aurora_notifier.android.gui.activities.main.fragments.aurora_factors;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.widget.LinearLayout;
import android.widget.TextView;

import se.gustavkarlsson.aurora_notifier.android.R;


public class AuroraFactorView extends LinearLayout {
	private TextView factorValueView;

	public AuroraFactorView(Context context) {
		super(context);
		init(context, null);
	}

	public AuroraFactorView(Context context, AttributeSet attrs) {
		super(context, attrs);
		init(context, attrs);
	}

	public AuroraFactorView(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		init(context, attrs);
	}

	@SuppressLint("SetTextI18n")
	private void init(Context context, AttributeSet attrs) {
		inflate(getContext(), R.layout.view_aurora_factor, this);
		if (attrs != null) {
			TypedArray auroraFactorViewAttributes = context.obtainStyledAttributes(attrs, R.styleable.AuroraFactorView);
			try {
				CharSequence name = auroraFactorViewAttributes.getText(R.styleable.AuroraFactorView_name);
				TextView nameView = (TextView) findViewById(R.id.aurora_factor_name);
				nameView.setText(name);
			} finally {
				auroraFactorViewAttributes.recycle();
			}
		}
		factorValueView = (TextView) findViewById(R.id.aurora_factor_value);
		// Used by IDE to display something
		if (isInEditMode()) {
			factorValueView.setText("value");
		}
	}

	public void setValue(String value) {
		factorValueView.setText(value);
	}
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/gui/views/AuroraFactorView.kt
package se.gustavkarlsson.skylight.android.gui.views

import android.content.Context
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.ShapeDrawable
import android.util.AttributeSet
import android.view.View
import android.widget.LinearLayout
import kotlinx.android.synthetic.main.view_aurora_factor.view.*
import se.gustavkarlsson.skylight.android.R
import se.gustavkarlsson.skylight.android.entities.Chance
import se.gustavkarlsson.skylight.android.util.ChanceToColorConverter


class AuroraFactorView : LinearLayout {

	private val chanceToColorConverter: ChanceToColorConverter

    constructor(context: Context) : super(context) {
		chanceToColorConverter = ChanceToColorConverter(context)
		init(context, null)
	}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
		chanceToColorConverter = ChanceToColorConverter(context)
		init(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
		chanceToColorConverter = ChanceToColorConverter(context)
		init(context, attrs)
    }

	private fun init(context: Context, attrs: AttributeSet?) {
        View.inflate(getContext(), R.layout.view_aurora_factor, this)
        if (attrs != null) {
            val auroraFactorViewAttributes = context.obtainStyledAttributes(attrs, R.styleable.AuroraFactorView)
			auroraFactorBadge.background.mutate()
            try {
                name = auroraFactorViewAttributes.getText(R.styleable.AuroraFactorView_name)
                badgeIcon = auroraFactorViewAttributes.getDrawable(R.styleable.AuroraFactorView_badgeIcon)
				val chanceFloat = auroraFactorViewAttributes.getFloat(R.styleable.AuroraFactorView_chance, Float.NaN)
				chance = if (chanceFloat.isNaN()) Chance.UNKNOWN else Chance(chanceFloat.toDouble())
				value = auroraFactorViewAttributes.getText(R.styleable.AuroraFactorView_value)
            } finally {
                auroraFactorViewAttributes.recycle()
            }
        }
        // Used by IDE to display something
        if (isInEditMode) {
			value = "value"
        }
    }

	var name: CharSequence?
		get() = auroraFactorName.text
		set(value) {
			auroraFactorName.text = value
		}

	var badgeIcon: Drawable?
		get() = auroraFactorBadge.drawable
		set(value) {
			auroraFactorBadge.setImageDrawable(value)
		}

	private var _chance = Chance.UNKNOWN
	var chance: Chance
		get() = _chance
		set(value) {
			val background = auroraFactorBadge.background
			val color = chanceToColorConverter.convert(value)
			when (background) {
				is ShapeDrawable -> background.paint.color = color
				is GradientDrawable -> background.setColor(color)
				is ColorDrawable -> background.color = color
			}
			_chance = value
		}

	var value: CharSequence?
		get() = auroraFactorValue.text
		set(value) {
			auroraFactorValue.text = value
		}
}
